base/steer/FairRun.cxx

```diff
@@ -79,14 +79,26 @@ void FairRun::AddTask(FairTask* t)
 {
   fTask->Add(t);
   fNTasks++;
+  StoreTaskNames(t);
+}
+//_____________________________________________________________________________
+void FairRun::StoreTaskNames(FairTask* t)
+{
+  if(t==NULL) return;
   fFileHeader->AddTaskClassName(t->ClassName());
+  if (t->GetListOfTasks()==NULL) return;
+  for(TObject* t2: *(t->GetListOfTasks()) )
+  {
+    StoreTaskNames((FairTask*)t2);
+  }
+  return;
```

Comment:
* Let's not call `GetListOfTasks()` twice.
* Downcast here with `static_cast` - safer than C-cast.

```cpp
const auto tasks = t->GetListOfTasks();
if(t)
{
  for(const auto task : *tasks)
￼   {
￼     StoreTaskNames(static_cast<FairTask*>(task));
￼   }
}
```